crates/rpc/src/method/get_storage_proof.rs

```diff
@@ -0,0 +1,784 @@
+        s.serialize_iter(
+            "contracts_storage_proofs",
+            self.contracts_storage_proofs.len(),
+            &mut self.contracts_storage_proofs.iter(),
+        )?;
+        s.serialize_field("global_roots", &self.global_roots)?;
+        s.end()
+    }
+}
+
+/// Returns all the necessary data to trustlessly verify:
+/// 1) Membership in the class trie.
+/// 2) Membership in the global state trie.
+/// 3) Membership in the contract storage trie.
+pub async fn get_storage_proof(context: RpcContext, input: Input) -> Result<Output, Error> {
+    const MAX_KEYS: usize = 100;
+    let mut total_keys = 0;
+
+    total_keys += input.class_hashes.as_ref().map_or(0, |hashes| hashes.len());
+    total_keys += input
+        .contract_addresses
+        .as_ref()
+        .map_or(0, |addresses| addresses.len());
+    total_keys += input.contracts_storage_keys.as_ref().map_or(0, |keys| {
+        keys.iter().map(|csk| csk.storage_keys.len()).sum::<usize>()
+    });
+
+    if total_keys > MAX_KEYS {
+        return Err(Error::ProofLimitExceeded {
+            limit: MAX_KEYS as u32,
+            requested: total_keys as u32,
+        });
+    }
+
+    let block_id = match input.block_id {
+        BlockId::Pending => {
+            // Getting proof of a pending block is not supported.
+            return Err(Error::ProofMissing);
+        }
+        other => other.try_into().expect("Only pending cast should fail"),
+    };
+
+    let span = tracing::Span::current();
+
+    let jh = tokio::task::spawn_blocking(move || {
+        let _g = span.enter();
+
+        let mut db = context
+            .storage
+            .connection()
+            .context("Opening database connection")?;
+
+        let tx = db.transaction().context("Creating database transaction")?;
+
+        // Use internal error to indicate that the process of querying for a particular
+        // block failed, which is not the same as being sure that the block is
+        // not in the db.
+        let header = tx
+            .block_header(block_id)
+            .context("Fetching block header")?
+            .ok_or(Error::BlockNotFound)?;
+
+        let class_root_idx = tx
+            .class_root_index(header.number)
+            .context("Querying class root index")?
+            .ok_or(Error::StorageProofNotSupported)?;
+
+        let class_root_hash = match tx
+            .class_trie_node_hash(class_root_idx)
+            .context("Querying class root hash")?
+        {
+            None if input.class_hashes.is_some() => return Err(Error::StorageProofNotSupported),
+            None => Felt::default(),
+            Some(hash) => hash,
+        };
+
+        let classes_proof = if let Some(class_hashes) = input.class_hashes {
+            let mut proofs = vec![];
+            for class_hash in class_hashes {
+                let proof =
+                    ClassCommitmentTree::get_proof(&tx, header.number, class_hash, class_root_idx)
```

Comment:
There's a low-hanging optimization possibility here: `ClassCommitmentTree` should probably handle multiple class hashes instead of just a single one. Right now we're loading the trie from storage for each and every class hash in the input -- while ideally we would just re-use the same trie (with all the nodes we've already loaded into memory).

This could be added in a follow-up PR of course.